documentation/api/query/v4/pql.markdown

```diff
@@ -0,0 +1,361 @@
+    facts { value < 4 }
+    facts { value <= 4 }
+    facts { value < 4 }
+
+The operator will only work on numbers however, any other types will return errors.
+
+#### Regexp: `~`
+
+For strings you can match using a regular expression pattern, by using the `~` operator and a valid
+regular expression:
+
+    nodes { certname ~ "foo.*" }
+
+* The regexp **must not** be surrounded by the slash characters (`/rexegp/`) that delimit regexps in many languages.
+* Every backslash character **must** be escaped with an additional backslash. Thus, a sequence like `\d` would be represented as `\\d`, and a literal backslash (represented in a regexp as a double-backslash `\\`) would be represented as a quadruple-backslash (`\\\\`).
+
+> **Note:** Regular expression matching is performed by the database
+> backend, so the available
+> [regexp features](http://www.postgresql.org/docs/9.4/static/functions-matching.html#POSIX-SYNTAX-DETAILS)
+> are determined by PostgreSQL.  For best results, use the simplest
+> and most common features that can accomplish your task.
+
+#### Array Match: `in`
+
+The `in` operator matches a field, or set of fields against either an array or a subquery.
+
+The `in` operator can be used in two ways. The simplest way is to see if a field, contains one of the values
+provided in a list of literal values:
+
+    nodes { certname in ['foo', 'bar', 'baz'] }
+
+The operator can also be used to ensure the values of a field match the fields returned, from a subquery,
+which has the form of a nested PQL query within the filter:
+
+    nodes { certname in facts[certname] { value = 'foo' } }
+
+With the subquery form, we can even match on multiple fields, as long as the fields being matched, and the
+subqueries projection fields match:
+
+    facts { [name, value] in facts[name, value] { value ~ 'foo' } }
+
+#### Null detection: `is null`, `is not null`
+
+Null values in PuppetDB are treated differently to other values. So to detect if a field is a null, instead of
+doing an exact match comparison, you must use either the `is null` or `is not null` operator.
+
+To test if a field contains a `null`:
+
+    nodes { deactivated is null }
+
+Or conversely, to test if a field does not contain a `null` value:
+
+    nodes { deactivated is not null }
+
+#### Regexp array match: `~>`
+
+The array matches using the regular expressions provided within in each element. Array indexes are coerced
+to strings.
+
+For example. the following query would query the path element, matching any ethernet mac address:
+
+    fact_contents { path ~> ["networking", "eth.*", "macaddress"] }
+
+### Boolean operators
+
+Boolean operators are used within PQL filters to join conditons together to perform the filtering
+test within PuppetDB.
+
+There are only 3 boolean operators today, in order of natural precedence:
+
+* `!` - performs a logical `not` or negation
+* `and` - performs a logical `and` or conjuction
+* `or` - performs a logical `or` or disjunction
```

Comment:
Good question! I thought about this, and I'm almost certain we don't want tokens that represent things, like fields and entities case-insensitive ...

So I think beyond the boolean operators: that leaves keywords like, group by, pagination stuff like limit, offset, order by etc. that could be case-insensitive. Also, there is functions.

I think at least whatever we choose it should be consistent. For me, I'm leaning towards case-sensitive, because it forces a style, however on the flip side, for SQL familiarity case-insensitive might be less surprising, but still - I feel like a firmer style decision isn't a bad thing.

One thing to note, leaving it case-sensitive, means we can change our minds easier later on, if we get demand. Its harder to revert case-insensitivity ...